controllers/servicefence_controller.go

```diff
@@ -313,26 +371,137 @@ func (r *ServicefenceReconciler) updateVisitedHostStatus(host *lazyloadv1alpha1.
 		}
 		return lazyloadv1alpha1.Destinations_ACTIVE
 	}
-
-	domains := make(map[string]*lazyloadv1alpha1.Destinations)
 	now := time.Now().Unix()
 
-	// update status of destinations of hosts with config
-	for h, strategy := range host.Spec.Host {
-		allHost := []string{h}
-		if hs := getDestination(h); len(hs) > 0 {
+	for h, strategy := range sf.Spec.Host {
+
+		fullHost := h
+		subDomains := strings.Split(h, ".")
+		switch len(subDomains) {
+		// full service name, like "reviews.default.svc.cluster.local", needs no action
+		case 5:
+		// short service name without namespace, like "reviews", needs to add namespace of servicefence and "svc.cluster.local"
+		case 1:
+			fullHost = fmt.Sprintf("%s.%s.svc.cluster.local", subDomains[0], sf.Namespace)
+		// short service name with namespace, like "reviews.default", needs to add "svc.cluster.local"
+		case 2:
+			fullHost = fmt.Sprintf("%s.%s.svc.cluster.local", subDomains[0], subDomains[1])
+		default:
+			log.Errorf("%s is invalid host, skip", h)
+			continue
+		}
+		if !isValidHost(fullHost) {
+			continue
+		}
+		if domains[fullHost] != nil {
+			continue
+		}
+
+		allHost := []string{fullHost}
+		if hs := getDestination(fullHost); len(hs) > 0 {
 			allHost = append(allHost, hs...)
 		}
 
-		domains[h] = &lazyloadv1alpha1.Destinations{
+		domains[fullHost] = &lazyloadv1alpha1.Destinations{
 			Hosts:  allHost,
 			Status: checkStatus(now, strategy),
 		}
 	}
+}
+
+// update domains with spec.namespaceSelector
+func addDomainsWithNamespaceSelector(domains map[string]*lazyloadv1alpha1.Destinations, sf *lazyloadv1alpha1.ServiceFence,
+	nsSvcCache map[string]map[string]struct{}, svcCache *sync.RWMutex) {
 
-	// update status with metric status
-	// -> (visited)host -> destinations
-	for metricName := range host.Status.MetricStatus {
+	svcCache.RLock()
+	defer svcCache.RUnlock()
+
+	for _, ns := range sf.Spec.NamespaceSelector {
+		svcs, ok := nsSvcCache[ns]
+		if !ok {
+			continue
+		}
+		for svc := range svcs {
+			subdomains := strings.Split(svc, "/")
+			fullHost := fmt.Sprintf("%s.%s.svc.cluster.local", subdomains[1], subdomains[0])
+			if !isValidHost(fullHost) {
+				continue
+			}
+			if domains[fullHost] != nil {
+				continue
+			}
+			allHost := []string{fullHost}
+			if hs := getDestination(fullHost); len(hs) > 0 {
+				allHost = append(allHost, hs...)
+			}
+			domains[fullHost] = &lazyloadv1alpha1.Destinations{
+				Hosts:  allHost,
+				Status: lazyloadv1alpha1.Destinations_ACTIVE,
+			}
+		}
+	}
+}
+
+// update domains with spec.labelSelector
+func addDomainsWithLabelSelector(domains map[string]*lazyloadv1alpha1.Destinations, sf *lazyloadv1alpha1.ServiceFence,
+	labelSvcCache map[string]map[string]struct{}, svcCache *sync.RWMutex) {
+
+	svcCache.RLock()
```

Comment:
I didn't mean to name the var as `svcCache` but do not pass a lock everywhere.
We can do it like this:

```go
type LabelSvcCache struct {
  data map[string]map[string]struct{}
  sync.RWMutex
}

// then we can:

lblSvcCache.RLock()  
// ...
```

Reply:
controllers/model.go
```go
type NsSvcCache struct {
	Data map[string]map[string]struct{}
	sync.RWMutex
}

type LabelSvcCache struct {
	Data map[LabelItem]map[string]struct{}
	sync.RWMutex
}